### DISCUSSION TITLE
MissingPluginException error on Firebase.initializeApp() calling when using FirebaseMessaging plugin in flutter tests

### DISCUSSION THREAD

**Discussion by OrlovPetr** [firebase/flutterfire#6613] [Q&A]
Hello!
Can you help me with this problem.
Trying to start this test:
```
void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  test('Test user instance', () async {
    /// Prepare user model for tests.
    await App().init();

    // expect(user.isAuth, true);
    // expect(user.miob, true);
    // expect(user.rememberMe, false);
  });
}
```

Method `App().init()` crashes on method `await InternalFirebase.whenAppStarted()` from `class InternalFirebase` which looks like this:
```
class InternalFirebase {
  static FirebaseMessaging _firebaseMessaging;

  static Future<void> whenAppStarted() async {
    await Firebase.initializeApp();

    _firebaseMessaging = FirebaseMessaging.instance;
    await _firebaseMessaging.requestPermission();
    await _firebaseMessaging.subscribeToTopic('APPS');
    await FirebaseNetwork(url: '${API_SERVICE_MAIN_URL}push/devices').registerDevice(_firebaseMessaging);
  }
}
```
I've been running tests in debug mode and crash is on this line: `await Firebase.initializeApp();`:
```
package:flutter/src/services/platform_channel.dart 156:7  MethodChannel._invokeMethod
MissingPluginException(No implementation found for method Firebase#initializeCore on channel plugins.flutter.io/firebase_core)
```

In pubspec.yaml I have this packages:
```
  firebase_messaging: ^10.0.3
  firebase_core: ^1.3.0
  firebase_analytics: ^8.1.2
```

Flutter doctor report:
```
[√] Flutter (Channel stable, 2.0.6, on Microsoft Windows [Version 10.0.19043.1110], locale ru-RU)
    • Flutter version 2.0.6 at C:\Flutter
    • Framework revision 1d9032c7e1 (3 months ago), 2021-04-29 17:37:58 -0700
    • Engine revision 05e680e202
    • Dart version 2.12.3

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:/Users/porlo/AppData/Local/Android/Sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = C:/Users/porlo/AppData/Local/Android/Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[√] Connected device (2 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 91.0.4472.124
    • Edge (web)   • edge   • web-javascript • Microsoft Edge 91.0.864.48

• No issues found!
```

This problem caused only on tests - in debug, profile and release modes all is ok.

I'm tried `flutter clean` and rebuild project, trying use hardcoded settings for` Firebase.initializeApp()` and looked to this issue: [https://github.com/FirebaseExtended/flutterfire/issues/3311](https://github.com/FirebaseExtended/flutterfire/issues/3311) but don't understand where is pb...

**Comment by felipecastrosales** [comment]
you can resolved this error?

**Reply by OrlovPetr** [reply]
I don't know, can be it resolved on Flutter 2.0.6 version, but I've upgraded to >3 version and all is ok. Didn't figure out what the problem was

**Reply by felipecastrosales** [reply]
> I don't know, can be it resolved on Flutter 2.0.6 version, but I've upgraded to >3 version and all is ok. Didn't figure out what the problem was

I'm having this issue in a UI test. Even using Flutter 3.3.4, the problem persists here.
Do you have any idea what I can do?

**Reply by OrlovPetr** [reply]
No, I hadn't this problem on Flutter 3.3.3.
All popular solutions doesn't help?

**Reply by felipecastrosales** [reply]
> No, I hadn't this problem on Flutter 3.3.3. All popular solutions doesn't help?

I'm having this issue when performing a widget test.
I don't know if for testing there is any additional configuration. Can you tell me?

**Reply by OrlovPetr** [reply]
As I know, for tests Firebase require mock configuration and don't work with real Firebase entity and API. Try to read about Flutter + Firebase test with mock. Maybe it ll help you...

**Reply by felipecastrosales** [reply]
> As I know, for tests Firebase require mock configuration and don't work with real Firebase entity and API. Try to read about Flutter + Firebase test with mock. Maybe it ll help you...

Wow, thank you so much for the guidance my friend, really.
In a few hours I will try and update you.

**Reply by OrlovPetr** [reply]
You are welcome)

**Reply by felipecastrosales** [reply]
@OrlovPetr I managed to take the test. 🥳 
In fact, what helped me the most was this.
* https://stackoverflow.com/a/64730015/13096514

There was also another mock that I hadn't configured properly and that was giving me inaccurate logs.
Thanks for your help and availability. 👍

**Reply by OrlovPetr** [reply]
Great! I'm glad you were able to solve your problem)
